test/test_autograd.py

```diff
@@ -2159,45 +2159,31 @@ def __del__(self):
         for _ in range(10):
             CollectOnDelete().forward(torch.randn(1, requires_grad=True)).backward()
 
-    # Delete this test when legacy custom autograd functions are deleted.
-    def test_naughty_legacy_variable_grad_fn(self):
-        class Id(Function):
-            def forward(self, x):
-                return x
-
-            def backward(self, grad_x):
-                return grad_x
-
-        self.assertRaises(RuntimeError, lambda: Variable(torch.zeros(1), _grad_fn=Id()))
-
-    # Delete this test when legacy custom autograd functions are deleted.
-    def test_naughty_legacy_function_backward_before_forward(self):
-        class Id(Function):
-            def forward(self, x):
-                return x
-
-            def backward(self, grad_x):
-                return grad_x
-
-        f = Id()
-        self.assertRaises(RuntimeError, lambda: f._do_backward((torch.zeros(0), ), False))
-
-    # Delete this test when legacy custom autograd functions are deleted.
-    def test_naughty_legacy_function_early_access(self):
+    def test_naughty_autograd_function_attr_access(self):
         class Id(Function):
+            @staticmethod
             def forward(self, x):
                 return x
 
+            @staticmethod
             def backward(self, grad_x):
                 return grad_x
-
         f = Id()
-        # A legacy autograd function is not fully initialized until you actually
-        # apply it.  That means a lot of accessors on them don't actually work.
-        # Test that we properly error in this case.
-        self.assertRaises(RuntimeError, lambda: f.register_hook(lambda x, y: None))
-        self.assertRaises(RuntimeError, lambda: f.next_functions)
-        self.assertRaises(RuntimeError, lambda: f.metadata)
+        # THPFunction is the base class of both grad_fn and autograd functions,
```

Comment:
Why can't we just forbid doing `Id()` altogether?
I'm sure that we can make sure that we can create the ctx internally without ever calling the `torch.autograd.Function.__init__`

Reply:
This might be risky because we never raised a warning for this particular code path. There is a probably a lot of code that does `instance = MyFunc(); instance.apply(args)`. I found several cases in test_autograd that does this.

Reply:
I've updated it to raise a warning when someone does `Id()` so we can remove it in a future version